### ISSUE TITLE
[1.1.1] Parallel installation issues on AIX (again)

### ISSUE THREAD

**Issue by makr** [openssl/openssl#8485]
I need to reopen the discussion from #6368 again. While things are generally working out fine if **no** LIBPATH environment variable is set, it gets hairy if you have multiple installations of, say libcrypto.a, appearing along your LIBPATH environment variable.

The actual situation we are facing right now is as follows:
* system provided OpenSSL 1.0.0 libraries in /usr/lib
* our own OpenSSL 1.1 libraries in /opt/.../openssl/lib
* among other paths, LIBPATH contains ...:/opt/.../openssl/lib:...:/usr/lib:/lib

The system's tools now no longer work, as the linker will find libcrypto.a in /opt/.../openssl/lib first, but it doesn't contain a libcrypto.so.1.0.0; our software works, however. When the order changes in LIBPATH, the set of working software changes, too.

So looking at how things seem to be done, at installation/update time we would either need to merge the system provided OpenSSL libraries into ours or the other way around. One could easily see that this is an unsupportable hairy construction. Especially if one or the other set of libraries gets updated ...

Does anyone have a hint on how to resolve this properly? Am I missing something? Should OpenSSL perhaps resort to not build and link the AIX way and not provide lib{crypto,ssl}.a with shared object members for 3rd-party software (via configure option maybe)?

**Comment by kaduk** [comment]
Would the "variant SONAMEs" of https://github.com/openssl/openssl/commit/822b5e2645a99bea15329bd66c9723c7e7119cdb work here?

**Comment by lzsiga** [comment]
Off: Well, AIX is very special with non-supporting version-numbers in shared libraries, but putting shared objects into *.a archives... On the plus side, the AIX!linker has a `-bipath` option that forces storing pathnames of shared objects.
To sum it up: I cannot offer a simple solutions, workarounde can be fabricated, here's some notes I've collected in the past decade regarding Aix: http://lzsiga.users.sourceforge.net/aix-linking.html

**Comment by makr** [comment]
> Would the "variant SONAMEs" of [822b5e2](https://github.com/openssl/openssl/commit/822b5e2645a99bea15329bd66c9723c7e7119cdb) work here?

No, AIX doesn't have a concept like this. Especially SONAME and versioned symbols are absent. However, one could build shared object like it is mentioned there, e.g. ```/opt/.../openssl/lib/libcrypto64.so.1.1```, but would then have to link to this library using its full path instead of just ```-lcrypto```.

@lzsiga thanks for the AIX linking FAQ, it is a really good read! The ```ipath``` option, however, is no solution as the software is delivered in binary form and can be installed anywhere in the system. This is necessary to support the installation of multiple versions of the same software at the same time on the same machine, e.g. for large migration projects.

**Comment by levitte** [comment]
(and people tell me VMS is strange...)

**Comment by lzsiga** [comment]
Hi, makr, as I said, I cannot offer complete working solution, so I don't think it would be decent to discuss ideas here. Perhaps on the AIX-forum of linuxquestions.org ?

**Comment by makr** [comment]
You might be right for the generic part of the question. However, the OpenSSL libraries are one of the few open source libraries that are also delivered on a system level, in /usr/lib, too. So I think it is worth an entry here, perhaps even with a conclusion, because people might stumble into this particular pitfall when they are building their own OpenSSL libraries on AIX.
BTW: The hint in your document about _not setting LIBPATH_ has triggered a thought about a possible solution to my issue. Instead of having an environment setup script that ensures the search path is set to find our libraries, I might need to deliver an environment setup script that removes the paths to our libraries as well as the standard paths from the search path ...

**Comment by lzsiga** [comment]
Oh, sorry, I've already started it here: https://www.linuxquestions.org/questions/aix-43/compiling-openssl-1-1-1b-on-aix-4175650442/

**Comment by makr** [comment]
After working for weeks now on an issue of a customer that has this problem as a root cause, I am calling out to the @OpenSSL team to revisit its decision about how to create dynamic libraries and other shared objects on AIX.

OpenSSL should go the other direction I mentioned initially in #6368 already. Instead of generating _libcrypto.a(libcrypto64.so.1.1)_ generate a _libcrypto64-1_1.so_. This way linking would just change into ```-lcrypto64-1_1```.
The other possibility is to generate a _libcrypto64.so.1.1_ and then linking changes to use absolute paths, however: ```-bnoipath -l/path/to/libcrypto64.so.1.1```. The binder's ```noipath``` option strips the path component from the dependency again.

Both ways would make it possible to have parallel installations of OpenSSL (with regard to API version that is, e.g. 1.0.0, 1.1 and perhaps 3.0). LIBPATH or LD_LIBRARY_PATH could be used, as the runtime linker would look for the correct version and not stop at the first libcrypto.a and complain if it doesn't contain the version looked for.

I would be willing to develop and provide a PR, if there is a general approval to move into this direction. I am also interested in the opinion of other users of AIX (esp. pinging @bfn34 because of #8788 ).

It is quite a journey :man_shrugging:

**Comment by lzsiga** [comment]
I don't think it would be good idea to reorganize things within an already released major version (ie 1.1.1).

For the next major release I'd suggest dropping the *.a-packaged shared objects, instead generate standalone shared objects like you do in Linux.

Mind you, it would be still problematic, as AIX doesn't have the concept of version numbers. (For details see the earlier linked document.)

**Comment by makr** [comment]
Any preference which of the two proposed options above you would go for?

**Comment by lzsiga** [comment]
I'd suggest dropping the '64' part from the file-name: `libssl.so.3.0.0.c` could be the actual filename, and `libssl.so.3.0` a symlink to it (or `libssl.so.3` or `libssl.so.3.0.0` depending on what the 'major version number' will be: 3.0 or 3 or 3.0.0)
At linkage, you specify libssl.so.3.0 libcrypto.so.3.0 so these strings will be encoded into the executable as dependencies.

**Comment by t8m** [comment]
@makr Is this still relevant?

**Comment by makr** [comment]
Well, there are still issues how the OpenSSL SOs are presented on AIX. However, 3.0 changed a lot compared to 1.1 and made things possible, that were not possible before. I have the occassional customer complaining about conflicts, but since 3.0 we've got it all resolved without stunts.
I close it as _obsolete_...
